### DISCUSSION TITLE
Whats the right way to detach a HWND from NativeControlHost?

### DISCUSSION THREAD

**Discussion by andy-westacott** [AvaloniaUI/Avalonia#15801] [Q&A]
Hi

I have a NativeControlHost containing a window.  This is a TopLevel window where I have messed with the window styles and rehosted it to the parent passed in CreateNativeControlCore().  All works great.

Now I need to detach this HWND again.  I reparent back to the windows original parent, re-apply the original styles.  All works great..

BUT I need to somehow tell the NativeControlHost to forget this HWND else once I remove the NativeControlHost from the visual tree.  The TopLevel window disappears, its not destroyed but seems to be reparented somehow by NativeControlHost to an instance of OffscreenParentWindow.  

Stepping through the windows with Spy++ I can see I do correctly reset its parent, its again a top level window at some point though OffscreenParentWindow (or similar) seems to grab the handle again and reparent.  Very odd.

Looking through the source its not immediately obvious what's going on but I do see this in Win32NativeControlHost.cs

```
class Win32NativeControlAttachment
{
            public void Dispose()
            {
                if (_child != null)
                    UnmanagedMethods.SetParent(_child.Handle, OffscreenParentWindow.Handle);
                _holder?.Dispose();
                _holder = null;
                _child = null;
                _attachedTo = null;
            }
```
Which I'm assuming is the culprit for reset the parent.

This is on Windows.
Any thoughts?

**Comment by maxkatz6** [comment]
The easiest would be to attach your HWND to the immediate parent, that you still have control of. And detach it on your request.

As an useful tip, you can do something like this with base.CreateNativeControlCore:
```
public override IPlatformHandle CreateNativeControlCore(IPlatformHandle parent)
{
    var defaultChild = base.CreateNativeControlCore(parent); // it will create another handle, that you can use to host your content, and detach from it dynamically.

     // TODO: create your HWND here and attach to the defaultChild.

    return defaultChild;
}
```

**Reply by andy-westacott** [reply]
Great.  Thank you, this was useful

Any tips on focus and embedded native controls.  What I'm seeing is I have focus on a native control and I hit tab and the focus, rather than move to the next focusable area in the native control, moves to the next focusable avalonia control
